candle-transformers/src/models/qwen3.rs

```diff
@@ -0,0 +1,441 @@
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Qwen3RmsNorm {
+    weight: Tensor,
+    eps: f64,
+}
+
+impl Qwen3RmsNorm {
+    fn new(dim: usize, eps: f64, vb: VarBuilder) -> Result<Self> {
+        let weight = vb.get((dim,), "weight")?;
+        Ok(Self { weight, eps })
+    }
+
+    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
+        let orig_dtype = xs.dtype();
+        let xs = xs.to_dtype(DType::F32)?;
+        let var = (xs.clone() * &xs)?.mean_keepdim(D::Minus1)?;
+        let rms = (var + self.eps)?.powf(-0.5)?;
+        let xs = xs.broadcast_mul(&rms)?;
+        let ws = self.weight.unsqueeze(0)?.unsqueeze(1)?;
+        xs.broadcast_mul(&ws)?.to_dtype(orig_dtype)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Qwen3HeadRmsNorm {
+    weight: Tensor,
+    eps: f64,
+}
+
+impl Qwen3HeadRmsNorm {
+    fn new(dim: usize, eps: f64, vb: VarBuilder) -> Result<Self> {
+        let weight = vb.get((dim,), "weight")?;
+        Ok(Self { weight, eps })
+    }
+
+    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
+        // xs: (B*L*H, D)
+        let orig_dtype = xs.dtype();
+        let xs = xs.to_dtype(DType::F32)?;
+        let var = (xs.clone() * &xs)?.mean_keepdim(D::Minus1)?;
+        let rms = (var + self.eps)?.powf(-0.5)?;
+        let xs = xs.broadcast_mul(&rms)?;
+        let ws = self.weight.unsqueeze(0)?;
+        xs.broadcast_mul(&ws)?.to_dtype(orig_dtype)
+    }
+}
+
+fn repeat_kv(kv: &Tensor, n_rep: usize) -> Result<Tensor> {
+    if n_rep == 1 {
+        return Ok(kv.clone());
+    }
+    let (b, h_kv, l, d) = kv.dims4()?;
+    kv.unsqueeze(2)?
+        .expand((b, h_kv, n_rep, l, d))?
+        .reshape((b, h_kv * n_rep, l, d))
+}
+
+#[derive(Debug, Clone)]
+struct Qwen3MLP {
+    gate_proj: Linear,
+    up_proj: Linear,
+    down_proj: Linear,
+    act_fn: Activation,
+}
+
+impl Qwen3MLP {
+    fn new(cfg: &Config, vb: VarBuilder) -> Result<Self> {
+        Ok(Self {
+            gate_proj: linear_no_bias(cfg.hidden_size, cfg.intermediate_size, vb.pp("gate_proj"))?,
+            up_proj: linear_no_bias(cfg.hidden_size, cfg.intermediate_size, vb.pp("up_proj"))?,
+            down_proj: linear_no_bias(cfg.intermediate_size, cfg.hidden_size, vb.pp("down_proj"))?,
+            act_fn: cfg.hidden_act,
+        })
+    }
+}
+
+impl Module for Qwen3MLP {
+    fn forward(&self, x: &Tensor) -> Result<Tensor> {
+        let lhs = x.apply(&self.gate_proj)?.apply(&self.act_fn)?;
+        let rhs = x.apply(&self.up_proj)?;
+        (lhs * rhs)?.apply(&self.down_proj)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Qwen3Attention {
+    // projections
+    q_proj: Linear,
+    k_proj: Linear,
+    v_proj: Linear,
+    o_proj: Linear,
+    // norms
+    q_norm: Qwen3HeadRmsNorm,
+    k_norm: Qwen3HeadRmsNorm,
+    // hyper params
+    num_heads: usize,
+    num_kv_heads: usize,
+    num_kv_groups: usize,
+    head_dim: usize,
+    hidden_size: usize,
+    // sliding window
+    _sliding_window: Option<usize>,
```

Comment:
Why is this prefixed with an `_`, is it unused at the moment, if so a comment should probably explain why it doesn't matter.

Reply:
Sliding window is null in all the configs I've seen on HF (like [here](https://huggingface.co/Qwen/Qwen3-8B/blob/main/config.json#L23)), But the option is still there. Not sure whether I should add a comment or just remove it.

Reply:
In that case, let's remove the `_` and have something like:
```
if sliding_window.is_some() {
    candle::bail!("sliding window is not suppored")
}
```
That way users will know if somehow they use a config with a sliding window.